Sources/HAP/Server/Device.swift

```diff
@@ -144,15 +230,107 @@ public class Device {
+
+        // Obtain new aid's for any accessories which don't already have one
+
+        self.accessories = [Accessory]()
+
+        // The first accessory must be aid 1
+        accessories[0].aid = 1
+        
+        addAccessories(accessories)
+    }
+    
+    public func addAccessories(_ newAccessories: [Accessory]) {
+ 
+        var configurationUpdated = false
+        
+        accessories.append(contentsOf: newAccessories)
+        
+        for accessory in newAccessories {
             accessory.device = self
-            accessory.aid = idGenerator.next()!
+            if (accessory.aid == 0) {
+                
+                // Check to see if the aid has been stored in the configuration data
+                
+                let dsuid = accessory.deviceSpecificUniqueId
+                if let aid = configuration.aidForAccessoryUniqueId[dsuid] {
+                    accessory.aid = aid
+                }
+            }
         }
+
+        for accessory in newAccessories {
+            // Verify that the aid is indeed unique
+            if accessory.aid != 0,
+               !isUniqueAID(accessory.aid, ignoring: accessory) {
+                logger.info("Accessory \(accessory.info.name.value ?? "unknown") has a duplicate accessory ID \(accessory.aid), replacing with fresh IID")
+                accessory.aid = 0
+            }
+            if (accessory.aid == 0) {
+                // Obtain a new aid, which has not already been used
+                repeat {
+                    accessory.aid = configuration.nextAID
+                } while (!isUniqueAID(accessory.aid, ignoring: accessory))
+                configurationUpdated = true
+
+                // Store the aid in the configuration data
+                
+                let dsuid = accessory.deviceSpecificUniqueId
+                configuration.aidForAccessoryUniqueId[dsuid] = accessory.aid
+            }
+        }
+        
+        // write configuration data to persist updated aid's and notify listeners
+        
+        if configurationUpdated {
+            updatedConfiguration()
+        }
+
+        
+    }
+    
+    public func removeAccessories(_ unwantedAccessories: [Accessory]) {
+        
+        var configurationUpdated = false
+        
+        for accessory in unwantedAccessories {
+            // Ensure the initial accessory is not removed, and that the accessory is in the list
+            if accessory.aid != 1 {
+                if let i = accessories.index(where: { $0 === accessory}) {
+                    accessories.remove(at: i)
+                    
+                    let dsuid = accessory.deviceSpecificUniqueId
+                    configuration.aidForAccessoryUniqueId.removeValue(forKey: dsuid)
+                    
+                    configurationUpdated = true
+                }
+            }
+        }
+        
+        // write configuration data to persist updated aid's
+        
+        if configurationUpdated {
```

Comment:
Calling this method should always update the configuration, as we eliminated most cases where we didn't;

* `unwantedAccessories` has invalid entries -> trap
* `unwantedAccessories` could be empty -> we could check beforehand, or maybe even also trap?

Reply:
OK